### ISSUE TITLE
Support for 2FA authentication

### ISSUE THREAD

**Issue by madchap** [matterhorn-chat/matterhorn#228]
Hi,

I know it is mentioned in the README.md that only user/pass is currently supported, but I could not find any outstanding issue asking for 2FA support.

Thank you.
Best,
fred

**Comment by jtdaugherty** [comment]
Thanks! We definitely need to take a look at this at some point. It is not on our current development plan (which runs through August) but we'd definitely be open to working with contributors if anyone wants to take a look at this.

**Comment by madchap** [comment]
Hi,

Just pinging... everytime I look back to Matterhorn, I am like "why am I not using this?". And it strikes me again :-( I hope you can get this prioritized somehow :+1:

**Comment by jtdaugherty** [comment]
@madchap - thanks for the ping. It's good to know that this would make Matterhorn useful to you!

To figure out when we might tackle this, we first need to investigate what would be required to do it, and in particular, whether it is even feasible (in case something about how it is implemented depends on the web client). We could at least make that investigation a priority in the coming months, but only after that's completed could we say anything definite about implementation. If someone else (such as yourself) wanted to do that, here's what we'd want to know:

* What is the protocol workflow for 2FA? What API requests need to be made between which entities, in what order?
* What are the user interface requirements to effectively support 2FA?
* Does the web browser need to be involved at all?
* For implementation, are there already helper libraries or would we have to implement new ones?

**Comment by madchap** [comment]
Pinging my friend @benoitknecht here, just in case :-P

**Comment by dylangerdaly** [comment]
If putting 2FA is a little too hard, would it be useful to allow for session 'injection'

Where you can specify the hostname, username and cookies and just log into an already active session?

**Comment by jtdaugherty** [comment]
@dylangerdaly no, that would not be advisable.

**Comment by jtdaugherty** [comment]
This has been migrated to the wiki: https://github.com/matterhorn-chat/matterhorn/wiki/Feature-Backlog

**Comment by xfalcox** [comment]
I got MFA working with a very ugly hack (long time without doing haskell)

https://github.com/xfalcox/matterhorn/commit/7d1aceeedc72890d2f01ce28cc61f2039e218ea7

**Comment by kquick** [comment]
Thanks for the update, @xfalcox !  Can you tell me why you think this is an ugly hack?  There are a few minor comments, but it looks like a pretty reasonable implementation at first glance.

It looks like it will generate the login prompt if the MFA is blank, and in my experience these are always updated on a frequent basis, so I wouldn't imagine that it's useful to get an MFA token from the config file.  I assume you are expected to use whatever authentication tool to get a current token and enter that into the login screen.  Am I right in these assumptions?

Do you know what happens if the user gets disconnected for long enough that the websocket connection to the server must be re-established, or if the local client's IP address changes?  Would it require re-authentication?

I'm also curious as to whether the login failure would indicate that an MFA token is needed: we could potentially arrange for a separate prompt to just obtain the MFA token, which would also accomodate any subsequent needs if it's also required in situations like the disconnects or IP updates I described above.

And lastly, are you using this on a regular basis, and if so, how well is it working for you?  Any areas where you think some additional work would be needed?

It would be great to explore some of the above and work towards having you submit a pull request to add this feature.

**Comment by jtdaugherty** [comment]
I also have some comments, which I'll leave as comments on the diffs. Thanks for your work, @xfalcox!

**Comment by jtdaugherty** [comment]
(My comments are largely aimed at changes that we'd want to see if this got merged at some point.)

**Comment by jtdaugherty** [comment]
I left a few comments in the patches. For me the biggest concern about this is how best to design the UI for it. Having a login field that lets you enter the normal username and password settings *and* an MFA token makes it hard to tell what will happen if I enter either, or neither. I think @kquick was thinking along these lines above, as well. I'd almost want to have a separate UI for this. That way, there's no confusion about what needs to be entered because you either enter all of the other authentication details *or* you enter a token. That would also be further helped by splitting the hostname and port fields out of the current authentication form, because they'd be used irrespective of the authentication method.

**Comment by jtdaugherty** [comment]
And for what it's worth, this discussion doesn't necessarily mean @xfalcox is on the hook to implement any discussed ideas; we just need to sort out an approach that we think will work. As for who implements it, @xfalcox or others can opt in to extend the patch.

**Comment by jtdaugherty** [comment]
I realized my comment above about a separate UI was actually way off - I was thinking of alternative token authentication, not second factors. So nevermind on that proposal. :)

**Comment by xfalcox** [comment]
> Thanks for the update, @xfalcox ! Can you tell me why you think this is an ugly hack? There are a few minor comments, but it looks like a pretty reasonable implementation at first glance.

Thanks for the kind words :blush:

> It looks like it will generate the login prompt if the MFA is blank, and in my experience these are always updated on a frequent basis, so I wouldn't imagine that it's useful to get an MFA token from the config file. I assume you are expected to use whatever authentication tool to get a current token and enter that into the login screen. Am I right in these assumptions?

Exactly, tokens live for 30 seconds so they can never live in a config file and must be input in the Login process.

Typical 2FA generations tools are: Google Authenticator, Authy and Yubikeys.

> Do you know what happens if the user gets disconnected for long enough that the websocket connection to the server must be re-established, or if the local client's IP address changes? Would it require re-authentication?

Judging by my the Mattermost web app and by my brief usage of libpurple mattermost implementation (which authenticates using the cookie) the mattermost cookies are long lived enough that being disconnected for a few minutes doesn't impact the session.

> I'm also curious as to whether the login failure would indicate that an MFA token is needed: we could potentially arrange for a separate prompt to just obtain the MFA token, which would also accomodate any subsequent needs if it's also required in situations like the disconnects or IP updates I described above.

That is how the web app does it. A valid auth for a user with 2FA enabled fails with 401 and a message tells that the token couldn't be validated. The webapp then asks for MFA and submits the whole login form again (username/password), this time passing the 2FA token as `token` in the JSON.

This buggy behavior is the whole base of my changes, since it showed me that getting MFA working was just a matter of one extra param to the same endpoint, making the implementation feasible for someone who isn't familiar with the codebase.

> And lastly, are you using this on a regular basis, and if so, how well is it working for you? Any areas where you think some additional work would be needed?

2FA is mandatory at my company so I couldn't test it until a few hours ago when I got it working. I'm now running it and it appears to be great! A couple of coworkers may also try it out as the electron mattermost "desktop" app isn't ideal for everyone.

> And for what it's worth, this discussion doesn't necessarily mean @xfalcox is on the hook to implement any discussed ideas; we just need to sort out an approach that we think will work. As for who implements it, @xfalcox or others can opt in to extend the patch.

I'd be happy for you guys to take my "proof of concept" and add it properly, mimicking the web app flow of:

1. Show login form
2. `POST` without token
3. Get 401, inspect reason, if the reason is bad token
4. Show 2FA form
5. `POST` a merged JSON between forms from 1 and 4

Is a bit too much for my level of familiarity with the codebase.

**Comment by chrysophylax** [comment]
Is there a branch or WIP: PR for this? I've too little background in Haskell to set this up safely :-)

**Comment by jtdaugherty** [comment]
@chrysophylax not that I know of.

**Comment by xfalcox** [comment]
@chrysophylax there is a release with MFA support at https://github.com/xfalcox/matterhorn/releases

**Comment by corbolais** [comment]
I'm all in for testing something released in this repo..

Something at least a bit confidence instilling..

Preferably bearing a title more like "`2FA implementation done`", rather than "`pre-release: super hacky MFA support`".

**Comment by xfalcox** [comment]
@corbolais if you don't want to trust the binary I uploaded, you can use my forks, review the change sets and compile it locally.

They are available at https://github.com/matterhorn-chat/matterhorn/issues/228#issuecomment-539816658

**Comment by corbolais** [comment]
@xfalcox Thank you.
I'm sorry, I really didn't want to offend. I should've written the comment differently. I am grateful, you implemented it that far. Otherwise there would be no such thing. Hopefully, the way to go is not too long now..

It's not about trust, it's about a clean solution, as you hinted at yourself.

FWIW, I won't use matterhorn any longer, see also https://github.com/mattermost/docs/issues/1006 and their extensive djungle of legalese with references to Terms of Use/Service, Rebranding, Whitelabeling etc.

Thanks again for your work, @xfalcox

**Comment by kquick** [comment]
@corbolais, would you please consider removing your third paragraph ("FWIW ...") comment above?  Matterhorn itself is clearly licensed and I believe your objections are with _Mattermost_ licensing, whilst Matterhorn is independently developed and not associated with the licensing issues you find objectionable.  You have posted your concerns directly on the Mattermost repo which is a more appropriate place for handling them, and don't believe those concerns are reflective of or directly related to the work on Matterhorn.

**Comment by jtdaugherty** [comment]
Thanks, @kquick - it's important for folks reading this comment thread to recognize the distinction between our project and the upstream project and the differences in licensing!

**Comment by msm-code** [comment]
Hi,

is there any work going on on this issue? In my company we decided to force 2FA for all employees. I also believe 2FA is very important for security. So I've hacked together this code, to keep using matterhorn. This is how login form looks after my changes:

![ss](https://user-images.githubusercontent.com/7026881/92124471-1e0f0b80-edfe-11ea-93c7-7fa4176e2dd6.png)

This works for me, but is ugly when user *don't* need 2FA token to login. My changes can be seen at PR #663.

I'm open for improving this PR, but looking for feedback first.

**Comment by jtdaugherty** [comment]
@msm-code Thank you for your work! We'll take a look. In the mean time, I'm curious: how do users come by their OTP values?

**Comment by jtdaugherty** [comment]
Here's a conversation I had with Harrison Healey, one of the server developers, on whether we can dynamically present a multi-factor auth token UI in response to an authentication failure:

> [11:19] jtdaugherty: Do mattermost servers respond with any kind of indication that two-factor auth is required? In other words, is the fact that 2fa is enabled detectable?
> [12:28] harrison: There's an endpoint that says whether or not the user has 2fa enabled after they've logged in
> [12:29] jtdaugherty: Okay. I'm looking for something that can be used before authentication in order to guide how the auth UI is presented.
> [12:30] harrison: That used to be available before logging in, but it was changed since it understandably leaked some data. I think the login method now has a specific error returned when 2fa is enabled
> [12:30] harrison: The way we've done it, the 2fa prompt comes after the username/password one. We try to log in with username/password, and then if that fails due to 2fa, we show the 2fa prompt. Then we use the previous value of the password instead of prompting the user for it again
> [12:33] jtdaugherty: Okay, thanks! That makes sense. Where can I find info on how the login API response will look in that situation?
> [12:34] harrison: I don't think we've generally documented error codes, but in this case, the returned error object will have the id `mfa.validate_token.authenticate.app_error`
> [12:36] jtdaugherty: Would you recommend against depending on that id? I'm wondering how likely it is that it would ever change if it is not part of a documented API.
> [12:37] harrison: You're totally fine to depend on that. It's had that value since we added MFA, and it would break the mobile app as well if we changed it
> [12:37] jtdaugherty: Okay. Thank you!

**Comment by msm-code** [comment]
> In the mean time, I'm curious: how do users come by their OTP values?

Depends on the user. Most people I know use Google Authenticator (an application on their phones). I use a three-line python script I've coded myself. There are also alternative applications, and hardware tokens for really paranoid.

>  [12:30] harrison: The way we've done it, the 2fa prompt comes after the username/password one. We try to log in with username/password, and then if that fails due to 2fa, we show the 2fa prompt. Then we use the previous value of the password instead of prompting the user for it again
> ...
 [12:37] harrison: You're totally fine to depend on that. It's had that value since we added MFA, and it would break the mobile app as well if we changed it

Great to hear that! I'll take a look at it then. The change so far was surprisingly easy, so hopefully I'll manage to implement that workflow without major problems.

**Comment by jtdaugherty** [comment]
@msm-code thanks! Bear in mind that @kquick and I will still be interested in the overall user experience of however this ends up being implemented, so please consider any more work on this iterative while we figure out what works best. If you're interested in continuing to hack on it then we would be grateful!

**Comment by jtdaugherty** [comment]
As for how users obtain OTP tokens, what I want us to be thinking about is how Matterhorn's UI should be designed to best accommodate how users obtain the tokens. I also want to avoid feature creep and provide the simplest possible approach that works for most users rather than trying to cover all cases. (A field you can paste the token into is probably the best we can do, although I could imagine providing a command config option similar to `tokencmd` in case users have a script that can obtain the token out of band.)

**Comment by chrysophylax** [comment]
Hi, has there been any more work on figuring out what works best for UX so this story can move forward?

**Comment by jtdaugherty** [comment]
@chrysophylax I'm afraid I and @kquick have not had any time to think about how to design the interface for this feature.

Thinking about it now, in case it helps anyone who's contributing:

Having an additional field for the OTP token may be a necessary way to get the token from the user, but showing it by default does not compose well with the existing authentication form, which is already getting to be unwieldy. In the latest version of Matterhorn, the form already has fields for the server URL (necessary), username/password (usually necessary), and access token (rarely used but still important). The conversation I had with one of the server developers (that I pasted above) suggests that if we attempt an authentication without including an OTP token, the server will respond with a particular error. That error could be used to then display a field to get the token from the user as needed. That would allow us to keep it out of the form by default and avoid the clutter.

**Comment by toastal** [comment]
>In the mean time, I'm curious: how do users come by their OTP values?

Just adding mine.

* For my phone: I scanned the QR code given to me by Mattermost into andOTP from F-Droid. To consume I can look at the screen.
* For my laptop: I inspected the image in the browser console. Sometimes it has a data attribute with the otpauth info, sometimes not. If it's a PNG like it was for Mattermost I download the image and if it's an SVG like some applications, I snip a screenshot of it. Running it through `zbarimg` it'll read the QR code and give you the otpauth info. That is then inserted into `pass` with the OTP extension which I can then `pass otp show ...`.
* For my hardware key: I use an OnlyKey and use the above info to create a new TOTP shortcut. To consume I just press a button on the key.

This gives me enough backups for convenience and redundancy in case of a device failure.

**Comment by jtdaugherty** [comment]
Thanks, @toastal!

**Comment by toastal** [comment]
I too am soon having gig where 2FA (TOTP) is a requirement, so after looking at it as an alternative to the Web UI, Matterhorn became a no-go. The admin doesn't have access tokens turned on either.

**Comment by msm-code** [comment]
Hi,

I've struggled with UI a bit, completely forgot about that PR and moved on with my life. I'm still interested in finishing my work and having it merged (right now I switch between my fork and web browser, depending on machine I'm using, which is far from good).

I'll have another go at improving my PR in the coming days. Sorry for taking so long.

**Comment by jtdaugherty** [comment]
Hi @msm-code - no problem, thanks for your update. I had posted a comment above where I had thrown out some ideas for how to manage the UI for this feature (using a particular server error response to prompt for a token). If you could investigate that, that would be very helpful. But there's no need to hurry to get this done - take your time!

**Comment by msm-code** [comment]
Pushed my changes! It only took a while (even though I had to learn how the UI framework works :thinking: ), sorry for dragging this on for so long.

Copy&paste from my commit message:

I've added a modal login window. The first stage looks as before:

![image](https://user-images.githubusercontent.com/7026881/146703077-85f17fd5-15c0-42f0-8bdc-c1aca7b886fd.png)

But after a successful login, OTP window shows up:

![image](https://user-images.githubusercontent.com/7026881/146703104-686bc05c-54e2-4e4b-afde-cd1c953ce258.png)

**Comment by msm-code** [comment]
Oh, also someone suggested that I should support otpcmd:

> also a otpcmd for the config.ini would be nice.

If possible, I'd like to submit this as a separate PR, because:

1. This one took me long enough. I don't want to delay it more.
2.Bare-bones support for OTP is way more important than QoL features like that (lack of MFA is dealbreaker when your organisation, like mine, enforces it).
3. I *personally* don't need it, so it's lower priority for me :)

But, if it's welcome, I'm happy to try and add this feature.

**Comment by toastal** [comment]
> I _personally_ don't need it, so it's lower priority for me

It's a higher priority to me so that I can keep the commands all behind password-store. Though it mostly just saves one fiddling step for my phone or running the command to the clipboard and pasting.

**Comment by msm-code** [comment]
Can you elaborate what `password-store` is? Is it a blocking issue for me?

What I mean is, with basic 2FA support Matterhorn is at least useable for everyone (my previous $dayjob and $ctfteam mattermosts require 2FA unconditionally). `otpcmd` sounds more like a QoL feature.

I also would like to have a ACK from maintainers before commiting to adding this feature (maybe they don't want it for some reason, or they would rather use environment variable, etc). Just so I don't have to reimplement it later.

But I'm happy to add `otpcmd`. My point was, ideally, I'd like to merge this PR first, and then submit another PR with `otpcmd`. Small and focused PRs usually work better :).

**Comment by jtdaugherty** [comment]
@msm-code Yes, let's hold off on a configuration setting in a separate change. I do think that it would be good to add that, though (called `otptokencmd` to distinguish it from the existing `tokencmd`).

**Comment by jtdaugherty** [comment]
The work for this has been merged. Thanks, @msm-code!
